core/src/local/persistent_state.rs

```diff
@@ -442,29 +442,56 @@ impl PersistentState {
 
     fn remove(&self, batch: &mut WriteBatch, r: &[DataType]) {
         let db = self.db.as_ref().unwrap();
-        let mut delete_cols: Vec<_> = self.indices
-            .iter()
-            .enumerate()
-            .map(|(i, index)| {
-                if self.primary_key.is_some() {
-                    (i + 1, &index.columns)
-                } else {
-                    (i, &index.columns)
+        let mut do_remove = move |indices: &[PersistentIndex], primary_key: &[u8]| {
+            // First delete the actual primary key row:
+            batch.delete(&primary_key).unwrap();
+
+            // Then delete any references that point _exactly_ to
+            // that row (i.e. not all matching rows):
+            for (i, index) in indices.iter().enumerate() {
+                // Construct a key with the index values, and serialize it with bincode:
+                let key = KeyType::from(index.columns.iter().map(|i| &r[*i]));
+                // Add 1 to i since we're slicing self.indices by 1..:
+                let prefix = Self::serialize_prefix((i + 1) as u32, &key);
+                for (raw_key, raw_value) in db.prefix_iterator(&prefix) {
+                    if &*raw_value == &primary_key[..] {
+                        batch.delete(&raw_key).unwrap();
+                    }
                 }
-            })
-            .collect();
+            }
+        };
 
+        // First retrieve the actual stored row that matches r:
         if let Some(ref pk_cols) = self.primary_key {
-            delete_cols.push((0, pk_cols));
-        }
+            let pk = KeyType::from(pk_cols.iter().map(|i| &r[*i]));
+            let prefix = Self::serialize_prefix(0, &pk);
+            let raw_value = db.get(&prefix).unwrap();
+            if let Some(raw) = raw_value {
+                let value: Vec<DataType> = bincode::deserialize(&*raw).unwrap();
+                if r != &value[..] {
```

Comment:
I think maybe this should `assert_eq!(r, &value[..])`, since we never expect get a negative that doesn't match a stored row.

Reply:
Good point. Added an `.expect()` for the case where the PK row isn't found as well. Not sure if there should be something similar for non-PK removals, e.g. where it can't find at least one row?

Reply:
In general, every removal should remove exactly one row. If they don't, it means we produced an incorrect negative, which should never happen.